runtimes/ocaml/runtime.ml

```diff
@@ -703,61 +722,60 @@ end
 
 let handle_default :
       'a.
-      source_position ->
+      source_position array ->
       (unit -> 'a) array ->
       (unit -> bool) ->
       (unit -> 'a) ->
       'a =
  fun pos exceptions just cons ->
-  let except =
-    Array.fold_left
-      (fun acc except ->
-        let new_val = try Some (except ()) with EmptyError -> None in
-        match acc, new_val with
-        | None, _ -> new_val
-        | Some _, None -> acc
-        | Some _, Some _ -> raise (ConflictError pos))
-      None exceptions
+  let len = Array.length exceptions in
+  let rec filt_except i =
+    if i < len then
+      match exceptions.(i) () with
+      | new_val -> (new_val, i) :: filt_except (i + 1)
+      | exception Empty -> filt_except (i + 1)
+    else []
   in
-  match except with
-  | Some x -> x
-  | None -> if just () then cons () else raise EmptyError
+  match filt_except 0 with
+  | [] -> if just () then cons () else raise Empty
+  | [(res, _)] -> res
+  | res -> error Conflict (List.map (fun (_, i) -> pos.(i)) res)
 
 let handle_default_opt
-    (pos : source_position)
+    (pos : source_position array)
     (exceptions : 'a Eoption.t array)
     (just : unit -> bool)
     (cons : unit -> 'a Eoption.t) : 'a Eoption.t =
-  let except =
-    Array.fold_left
-      (fun acc except ->
-        match acc, except with
-        | Eoption.ENone _, _ -> except
-        | Eoption.ESome _, Eoption.ENone _ -> acc
-        | Eoption.ESome _, Eoption.ESome _ -> raise (ConflictError pos))
-      (Eoption.ENone ()) exceptions
+  let len = Array.length exceptions in
+  let rec filt_except i =
+    if i < len then
+      match exceptions.(i) with
+      | Eoption.ESome _ as new_val -> (new_val, i) :: filt_except (i + 1)
+      | Eoption.ENone () -> filt_except (i + 1)
+    else []
   in
-  match except with
-  | Eoption.ESome _ -> except
-  | Eoption.ENone _ -> if just () then cons () else Eoption.ENone ()
-
-let no_input : unit -> 'a = fun _ -> raise EmptyError
+  match filt_except 0 with
+  | [] -> if just () then cons () else Eoption.ENone ()
+  | [(res, _)] -> res
+  | res -> error Conflict (List.map (fun (_, i) -> pos.(i)) res)
 
 (* TODO: add a compare built-in to dates_calc. At the moment this fails on e.g.
    [3 months, 4 months] *)
-let compare_periods (p1 : duration) (p2 : duration) : int =
+let compare_periods pos (p1 : duration) (p2 : duration) : int =
   try
     let p1_days = Dates_calc.Dates.period_to_days p1 in
     let p2_days = Dates_calc.Dates.period_to_days p2 in
     compare p1_days p2_days
-  with Dates_calc.Dates.AmbiguousComputation -> raise UncomparableDurations
+  with Dates_calc.Dates.AmbiguousComputation ->
```

Comment:
Re the comment I made earlier : this raises a runtime exception, not a `Dates_calc` exception, so why catch a `Dates_calc` exception ?

Reply:
Exactly, here we are in the runtime so we catch the exception from `Dates_calc` to turn it into a proper `Runtime` exception ; the function in the earlier example actually had the comment:

```ocaml
(* This could fail, but is expected to only be called with known, already
   validated arguments by the generated code *)
```

since it would be called by the compiler, but is outside of what is expected to fail in the runtime (the compiled code *can* call it, but only on values already checked at compile-time)
I guess what led me to this was the intent to avoid catching `Error`s from the runtime.

Reply:
(in other words and to clarify my thoughts: `Runtime.date_of_numbers` failing _cannot_ be a runtime error, so it shouldn't belong to the type `Runtime.errors`)